### PR TITLE
Development/event queue

### PR DESCRIPTION
**Pull request by Dyronix**

### CODE REVIEW

File: source/1_engine/rex_engine/src/event_system.cpp [hunk 1]
```diff
@@ -14,18 +14,75 @@ namespace rex
...
+            ,m_front()
+            ,m_back()
+            ,m_current(&m_front)
+            ,m_next(&m_back)
+        {
+        }
+
+        void subscribe(EventType type, const EventFunction& function)
+        {
+          m_subscriptions[type].push_back(function);
```
Comment (RisingLiberty) [inline source/1_engine/rex_engine/src/event_system.cpp:33]: just to note, this is not thread save, as in, the entire class is not thread safe

File: source/1_engine/rex_engine/src/event_system.cpp [hunk 2]
```diff
@@ -14,18 +14,75 @@ namespace rex
...
+              delegate(evt);
+            }
+          }
+
+          m_current->clear();
+        }
+
+        void present()
+        {
+          rsl::vector<Event>* temp = m_current;
```
Comment (RisingLiberty) [inline source/1_engine/rex_engine/src/event_system.cpp:58]: why not use rsl::swap?

**Review by RisingLiberty — CHANGES_REQUESTED** [review]

File: source/1_engine/rex_engine/src/event_system.cpp (hunk 1 above)
Comment (Dyronix) [reply source/1_engine/rex_engine/src/event_system.cpp:33]: I think there are a lot more functionalities within REX that are not thread-safe. As we are single-threaded for now I think it's fine.

File: source/1_engine/rex_engine/src/event_system.cpp (hunk 2 above)
Comment (Dyronix) [reply source/1_engine/rex_engine/src/event_system.cpp:58]: Dunno, it does the same thing just within a function call.
I'll adjust it if you want.

Comment (RisingLiberty) [reply source/1_engine/rex_engine/src/event_system.cpp:58]: that's why we have functions, to wrap logic so I'd recommend to use it here

**Comment by RisingLiberty** [comment]
this is to resolve https://github.com/New-Creta/Rex/issues/202

**Comment by RisingLiberty** [comment]
I updated the code, it's ready for review

**Comment by RisingLiberty** [comment]
I think the best way to have this have multiple producer single consumer kinda thing.
multiple threads can push events but only 1 thread processes them.
that's currently not implemented because we don't have a thread pool/job system yet

**Review by RisingLiberty — APPROVED** [review]

**Pull request merged**